### PR TITLE
Controller pickup

### LINKED ISSUE #191: Pick up and release
Pick a selected object up and release it when pressing some button.

### PR DESCRIPTION
**Pull request by AncientFungus**
(Pull Request made by Jens Lomander)

Things can now be picked up by controller in such a way that an object's parent changes to the corresponding controller when you press a certain button. For now each thing you pick up is hard-coded to get pickup, but once the trigger system is in, we'll be able to much more generally pick up whatever our controller is intersecting with.

Will create a new issue for that the integration with triggers.
fixes #191

### CODE REVIEW

File: src/Engine/Component/VRDevice.cpp [hunk 1]
```diff
@@ -5,7 +5,6 @@
 
 namespace Component {
     VRDevice::VRDevice() {
```
Comment (NinjaDanz3r) [inline src/Engine/Component/VRDevice.cpp]: MINOR Member variable 'VRDevice::controllerID' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L582](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/770d646da7c83d99d9a62b5fb594a8706613c0eb/src/Engine/Manager/ScriptManager.cpp#L582): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L846](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/770d646da7c83d99d9a62b5fb594a8706613c0eb/src/Engine/Manager/ScriptManager.cpp#L846): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L852](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/770d646da7c83d99d9a62b5fb594a8706613c0eb/src/Engine/Manager/ScriptManager.cpp#L852): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L860](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/770d646da7c83d99d9a62b5fb594a8706613c0eb/src/Engine/Manager/ScriptManager.cpp#L860): The function 'ExecuteScriptMethod' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [VRManager.cpp#L181](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/770d646da7c83d99d9a62b5fb594a8706613c0eb/src/Engine/Manager/VRManager.cpp#L181): The function 'GetVRDevices' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Manager/ScriptManager.cpp [hunk 2]
```diff
@@ -84,7 +85,8 @@ void AngelScriptDebugLineCallback(asIScriptContext *ctx, const std::map<std::str
 }
 
 void print(const std::string& message) {
-    Log() << message;
+    std::cout << message;
```
Comment (Chainsawkitten) [inline src/Engine/Manager/ScriptManager.cpp]: Revert this change (this should be changed in the logging tool instead. You don't have to include iostream then.

File: src/Engine/Manager/ScriptManager.cpp [hunk 3]
```diff
@@ -115,14 +117,14 @@ glm::vec2 GetCursorXY() {
...
-    Managers().scriptManager->SendMessage(recipient, type);
+    Managers().scriptManager->SendMessage(recipient, Managers().scriptManager->currentEntity, type);
 }
 
 void RestartScene() {
     Hymn().restart = true;
 }
 
-bool IsIntersect(Entity* checker, Entity* camera) {
+Entity* IsIntersect(Entity* checker) {/*
```
Comment (Chainsawkitten) [inline src/Engine/Manager/ScriptManager.cpp]: Is this intentional or was it changed and commented out for testing purposes?

File: src/Engine/Manager/ScriptManager.hpp [hunk 4]
```diff
@@ -143,10 +143,13 @@ class ScriptManager {
         
         /// The entity currently being executed.
         Entity* currentEntity;
+
+        std::vector<Entity*> ScriptManager::GetUpdateEntities();
```
Comment (Chainsawkitten) [inline src/Engine/Manager/ScriptManager.hpp]: Undocumented public method.

File: src/Engine/Manager/VRManager.cpp [hunk 5]
```diff
@@ -128,15 +128,34 @@ void VRManager::SetScale(float scale) {
...
+                vr::VRControllerState_t controllerState;
+                if (vrSystem->GetControllerState(unDevice, &controllerState, sizeof(controllerState))) {
+                    pressedTrackedDevice[unDevice] = controllerState.ulButtonPressed == 0;
+                    if (controllerState.ulButtonPressed & vr::ButtonMaskFromId(buttonID))
+                        return true;
+                }
+            }
         }
     }
+    else if (ID == 2) {
```
Comment (Chainsawkitten) [inline src/Engine/Manager/VRManager.cpp]: else on same line as bracket

File: Game/Hymn.json [hunk 6]
```diff
@@ -32,20 +32,4116 @@
...
+			"size" : 3
 		}
 	],
 	"scripts" : 
 	[
 		"Resources/Opening Scene/Lever/Lever",
 		"Resources/Opening Scene/Gate/Gates",
 		"Resources/Constant Scene/Minecart/Script/Minecart",
-		"Resources/Constant Scene/Minecart/Script/Player",
+		"Resources/Constant Scene/Minecart/Script/Camera",
```
Comment (Chainsawkitten) [inline Game/Hymn.json]: Remove tons of duplicate scripts

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L582](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/db3aedd68e0dbf2f7cf08c37be1b608fc14d776f/src/Engine/Manager/ScriptManager.cpp#L582): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L846](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/db3aedd68e0dbf2f7cf08c37be1b608fc14d776f/src/Engine/Manager/ScriptManager.cpp#L846): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L852](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/db3aedd68e0dbf2f7cf08c37be1b608fc14d776f/src/Engine/Manager/ScriptManager.cpp#L852): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L860](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/db3aedd68e0dbf2f7cf08c37be1b608fc14d776f/src/Engine/Manager/ScriptManager.cpp#L860): The function 'ExecuteScriptMethod' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [VRManager.cpp#L181](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/db3aedd68e0dbf2f7cf08c37be1b608fc14d776f/src/Engine/Manager/VRManager.cpp#L181): The function 'GetVRDevices' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L580](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40f9572fc4ec316757736e7d96421f79c0372237/src/Engine/Manager/ScriptManager.cpp#L580): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L844](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40f9572fc4ec316757736e7d96421f79c0372237/src/Engine/Manager/ScriptManager.cpp#L844): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L850](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40f9572fc4ec316757736e7d96421f79c0372237/src/Engine/Manager/ScriptManager.cpp#L850): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L858](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40f9572fc4ec316757736e7d96421f79c0372237/src/Engine/Manager/ScriptManager.cpp#L858): The function 'ExecuteScriptMethod' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [VRManager.cpp#L181](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40f9572fc4ec316757736e7d96421f79c0372237/src/Engine/Manager/VRManager.cpp#L181): The function 'GetVRDevices' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L580](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/5f792163f4e6605e3e1c5e2e615da269de40ff22/src/Engine/Manager/ScriptManager.cpp#L580): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L844](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/5f792163f4e6605e3e1c5e2e615da269de40ff22/src/Engine/Manager/ScriptManager.cpp#L844): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L850](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/5f792163f4e6605e3e1c5e2e615da269de40ff22/src/Engine/Manager/ScriptManager.cpp#L850): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L858](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/5f792163f4e6605e3e1c5e2e615da269de40ff22/src/Engine/Manager/ScriptManager.cpp#L858): The function 'ExecuteScriptMethod' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [VRManager.cpp#L180](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/5f792163f4e6605e3e1c5e2e615da269de40ff22/src/Engine/Manager/VRManager.cpp#L180): The function 'GetVRDevices' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Manager/VRManager.cpp (hunk 5 above)
Comment (andens) [reply src/Engine/Manager/VRManager.cpp]: Actually, this entire else block is mostly a duplicate of the one above. We could store the desired role enum depending on `ID` and then check against that variable instead of hardcoding values. This way we can have one block that handles both controllers.

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L580](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/7addb986f8ec8e62f292c44f096dde44a9d9acb9/src/Engine/Manager/ScriptManager.cpp#L580): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L844](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/7addb986f8ec8e62f292c44f096dde44a9d9acb9/src/Engine/Manager/ScriptManager.cpp#L844): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L850](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/7addb986f8ec8e62f292c44f096dde44a9d9acb9/src/Engine/Manager/ScriptManager.cpp#L850): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L858](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/7addb986f8ec8e62f292c44f096dde44a9d9acb9/src/Engine/Manager/ScriptManager.cpp#L858): The function 'ExecuteScriptMethod' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [VRManager.cpp#L169](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/7addb986f8ec8e62f292c44f096dde44a9d9acb9/src/Engine/Manager/VRManager.cpp#L169): The function 'GetVRDevices' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'
